File: client/src/main/js/components/Rules.js
import React from 'react';
import {DeadlineRule} from './DeadlineRule';
import {Button, FormGroup} from 'reactstrap';
import {Api} from '../Api';
import {Notification} from './Notification';
import {credentials} from '../scripts';
import ReactDOM from 'react-dom';

export class Rules extends React.Component {

    constructor(props) {
        super(props);
        this.onDeadlineChange = this.onDeadlineChange.bind(this);
        this.submitRulesChanges = this.submitRulesChanges.bind(this);

        this.state = {
            deadline: props.task.deadline,
            saveDisabled: true,
        }
    }

    onDeadlineChange(newDeadline) {
        this.setState({deadline: newDeadline, saveDisabled: false});
    }

    submitRulesChanges() {
        Api.updateRules(credentials(),
            this.props.course.name,
            this.props.task.branch,
            this.state.deadline,
            () => ReactDOM.render(
                <Notification succeed message={`${this.props.task.branch} rules was updated!`}/>,
                document.getElementById('notifications')
            ),
            response => ReactDOM.render(
                <Notification message={`${this.props.task.branch} rules was not updated due to: ${response}`}/>,
                document.getElementById('notifications')
            )
        )
    }

    render() {
        return (
            <section className="rules-list">
                <DeadlineRule task={this.props.task} onDeadlineChange={this.onDeadlineChange}/>
                <FormGroup>
                    <Button color="primary"
                            onClick={this.submitRulesChanges}
                            disabled={this.state.saveDisabled ? 'disabled' : ''}
                    >Save rules</Button>
                </FormGroup>
            </section>
        )
    }
}

File: frontend/src/main/kotlin/org/flaxo/frontend/component/Rules.kt
package org.flaxo.frontend.component

import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import kotlinx.html.js.onClickFunction
import org.flaxo.common.data.Course
import org.flaxo.common.data.DateTime
import org.flaxo.common.data.Task
import org.flaxo.frontend.Container
import org.flaxo.frontend.Notifications
import org.flaxo.frontend.OnTaskChange
import org.flaxo.frontend.client.FlaxoClient
import org.flaxo.frontend.client.FlaxoHttpException
import org.flaxo.frontend.credentials
import react.RBuilder
import react.RComponent
import react.RProps
import react.RState
import react.dom.button
import react.dom.div
import react.setState

/**
 * Adds task rules menu.
 */
fun RBuilder.rules(course: Course, task: Task, onUpdate: OnTaskChange) = child(Rules::class) {
    attrs {
        this.course = course
        this.task = task
        this.onUpdate = onUpdate
    }
}

private class RulesProps(var course: Course, var task: Task, var onUpdate: OnTaskChange) : RProps

private class RulesState(var deadline: DateTime?) : RState

private class Rules(props: RulesProps) : RComponent<RulesProps, RulesState>(props) {

    private val flaxoClient: FlaxoClient

    init {
        flaxoClient = Container.flaxoClient
        state.deadline = props.task.deadline
    }

    override fun RBuilder.render() {
        div {
            deadlineRule(props.task) { newDeadline -> setState { this.deadline = newDeadline } }
            button(classes = "btn btn-primary") {
                attrs {
                    onClickFunction = { GlobalScope.launch { submitRulesChanges() } }
                    disabled = state.deadline == props.task.deadline
                }
                +"Update rules"
            }
        }
    }

    private suspend fun submitRulesChanges() {
        credentials?.also { credentials ->
            try {
                val updatedTask = flaxoClient.updateRules(credentials, props.course.name, props.task.branch,
                        state.deadline?.toDateString())
                props.onUpdate(updatedTask)
                Notifications.success("Task rules has been updated.")
            } catch (e: FlaxoHttpException) {
                console.log(e)
                Notifications.error("Error occurred while updating tasks rules.", e)
            }
        }
    }
}
